### PR TITLE
fix(typescript): fix types and improve basic test, add router

### PR DESCRIPTION
**Pull request by gaspard**
With these fixes, my ts app compiles without issues.

I had to add @types/react so that type verification works correctly.

I haven't dived into the details of function-tree types but they seem weird to me. We really need to have someone using TS with latest function-tree to validate them.

@reflog what is your current use of these types in ft ?

### CODE REVIEW

File: packages/node_modules/cerebral/devtools.d.ts [hunk 1]
```diff
@@ -1,13 +1,17 @@
...
-  allowedTypes?: any[]
-  doReconnect?: boolean
+  reconnect?: boolean
+  reconnectInterval?: number
+  storeMutations?: boolean
 }
 
-export default class DevTools {
-  constructor(options: DevOptions);
+export interface DevToolsInterface {
```
Comment (Guria) [inline packages/node_modules/cerebral/devtools.d.ts]: what this for? it looks very strange for me

**Comment by Guria** [comment]
hmm. @gaspard I am having issues with running `npm test` locally from monorepo root. Some packages tests fails. But at the same time `npm test` from package folder works fine. But at the same time there is no such problem on travis. I am really confused here.

File: packages/node_modules/@cerebral/router/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+export interface StringRoutes {
```
Comment (Guria) [inline packages/node_modules/@cerebral/router/index.d.ts]: let's not export interfaces and types until it really necessary

**Comment by Guria** [comment]
I have added a commit showing how we can avoid exporting strange opaque interfaces.
Lets scan across all typings and get rid of unnecessary exports.

File: packages/node_modules/@cerebral/router/index.d.ts (hunk 2 above)
Comment (gaspard) [inline packages/node_modules/@cerebral/router/index.d.ts]: My TypeScript app uses routes, how should it work without declaring them ?

**Comment by gaspard** [comment]
@Guria I looked at your changes: this is great. Thanks. Now I'll just add a fix for the timeout issue and we should be all good, at least for this first phase.

**Comment by gaspard** [comment]
@Guria Did you have a look at FunctionTree types ? I don't understand them.

File: packages/node_modules/cerebral/package.json [hunk 3]
```diff
@@ -23,7 +23,7 @@
   },
   "scripts": {
     "test": "mocha --compilers js:babel-register \"src/**/*.test.js\" && npm run test_ts",
-    "test_ts": "mocha --timeout 5000 --compilers js:babel-register \"typescript/typescript.spec.js\"",
+    "test_ts": "mocha -r \"typescript/test.setup.js\" --compilers js:babel-register \"typescript/typescript.spec.js\"",
```
Comment (henri-hulski) [inline packages/node_modules/cerebral/package.json]: I think this should also be changed in `function-tree` and `@cerebral/http`.

**Comment by henri-hulski** [comment]
Hmm I think @Guria meant his changes as an example which should be adopted for all unnecessary exports.

**Comment by henri-hulski** [comment]
When running `npm test` locally in 'packages/node_modules/cerebral' directory I get the following error output:
```
  1) TypeScript definitions Uncaught error outside test suite:
     Uncaught AssertionError: Semantic: /home/pacha/web/cerebral/cerebral/packages/node_modules/cerebral/react.d.ts (1
,24): Could not find a declaration file for module 'react'. '/home/pacha/web/cerebral/cerebral/node_modules/react/reac
t.js' implicitly has an 'any' type.
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:17:15
      at Array.forEach (native)
      at handleDiagnostics (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:
14:17)
      at compile (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:27:5)
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:55:13
      at next (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:74:24)
￼
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:87:21
      at FSReqWrap.oncomplete (fs.js:123:15)
 
  2) TypeScript definitions should compile against index.d.ts:
     Uncaught AssertionError: Semantic: /home/pacha/web/cerebral/cerebral/packages/node_modules/cerebral/react.d.ts (1
,24): Could not find a declaration file for module 'react'. '/home/pacha/web/cerebral/cerebral/node_modules/react/reac
t.js' implicitly has an 'any' type.
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:17:15
      at Array.forEach (native)
      at handleDiagnostics (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:
14:17)
      at compile (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:27:5)
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:55:13
      at next (/home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:74:24)
      at /home/pacha/web/cerebral/cerebral/node_modules/typescript-definition-tester/dist/index.js:87:21
      at FSReqWrap.oncomplete (fs.js:123:15)
```
The same when running `npm test` in root directory (of course fixts branch).

File: packages/node_modules/@cerebral/http/test/setup.js [hunk 4]
```diff
@@ -0,0 +1 @@
+// noop
```
Comment (henri-hulski) [inline packages/node_modules/@cerebral/http/test/setup.js]: Should this file not be removed?

File: packages/node_modules/cerebral/test/setup.js [hunk 5]
```diff
@@ -0,0 +1 @@
+// noop
```
Comment (henri-hulski) [inline packages/node_modules/cerebral/test/setup.js]: Same here.

File: packages/node_modules/function-tree/test/setup.js [hunk 6]
```diff
@@ -0,0 +1 @@
+// noop
```
Comment (henri-hulski) [inline packages/node_modules/function-tree/test/setup.js]: And here.

File: packages/node_modules/@cerebral/router/index.d.ts (hunk 2 above)
Comment (Guria) [inline packages/node_modules/@cerebral/router/index.d.ts]: can you show example? are all exports here is really necessary?

Comment (Guria) [inline packages/node_modules/@cerebral/router/index.d.ts]: router.ts file doesn't uses any of this exports.

Comment (gaspard) [inline packages/node_modules/@cerebral/router/index.d.ts]: We decided not to make these exports a public API so I thought they should not go into the `router.ts` file.

Here is my current router definition:

```js
const router =
Router
( { routes:
    [ { path: '/'
      , map:
        { score: compute
          ( state`score`, prepareUrlParams )
        }
      , rmap:
        { 'score': compute 
          ( props`score`
          , ( opts: Partial<ScoreUrlType>
            ) => {
              if ( ! opts ) return baseScoreState
              return validateUrlParams ( baseScoreState, opts )
            }
          )
        }
      }
    ]
    , query: true
    , onlyHash: true
  }
)
```

**Comment by gaspard** [comment]
Rebased against latest master branch.

File: packages/node_modules/@cerebral/router/index.d.ts (hunk 2 above)
Comment (gaspard) [inline packages/node_modules/@cerebral/router/index.d.ts]: I will remove `StringRoutes`. This is legacy compatibility.

**Pull request merged**